src/test/java/bio/terra/app/controller/SearchApiControllerTest.java

```diff
@@ -0,0 +1,67 @@
+package bio.terra.app.controller;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.verify;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+import bio.terra.common.category.Unit;
+import bio.terra.service.iam.IamAction;
+import bio.terra.service.iam.IamResourceType;
+import bio.terra.service.iam.IamService;
+import bio.terra.service.search.SnapshotSearchMetadataDao;
+import java.util.UUID;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.http.MediaType;
+import org.springframework.test.context.junit4.SpringRunner;
+import org.springframework.test.web.servlet.MockMvc;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(properties = "features.search.api=enabled")
+@AutoConfigureMockMvc
+@Category(Unit.class)
+public class SearchApiControllerTest {
+
+  @Autowired private MockMvc mvc;
+
+  @MockBean private IamService iamService;
+
+  @MockBean private SnapshotSearchMetadataDao snapshotMetadataDao;
+
+  private static final String ENDPOINT = "/api/repository/v1/search/{id}/metadata";
+
+  @Test
+  public void testUpsert() throws Exception {
+    var id = UUID.randomUUID();
+    var json = "{\"actual\": \"data\"}\"}";
+    mvc.perform(put(ENDPOINT, id).contentType(MediaType.APPLICATION_JSON).content(json))
+        .andExpect(status().isOk());
+    verifyAuthApi(id);
+    verify(snapshotMetadataDao).putMetadata(id, json);
+  }
+
+  @Test
+  public void testDelete() throws Exception {
+    var id = UUID.randomUUID();
+    mvc.perform(delete(ENDPOINT, id)).andExpect(status().isNoContent());
+    verifyAuthApi(id);
+    verify(snapshotMetadataDao).deleteMetadata(id);
+  }
+
+  private void verifyAuthApi(UUID id) {
```

Comment:
How is this expected to work with a random uuid?

Reply:
It verifies that the `iamService` API was called with the uuid provided. So the uuid is random for each test run but is the same between the call to the endpoint (via `mvc.perform()`) and the iam service being verified.

Reply:
To answer the question about how does the code not fail because I'm not mocking `when(snapshotMetadataDao).putMetadata()` etc using the random uuid, it's not necessary because the mocked DAO APIs are `void`, so the calling code isn't dependent on the results. Mockito records all calls to the mocks, so the test can use `verify()` to find out if the mock was called with the arguments test expects.